File: ghost-sdk/tools/openlog/open_log.m
%{
 * Copyright (C) Ghost Robotics - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Avik De <avik@ghostrobotics.io> and Pranav Bhounsule
%}
function s = open_log(fname)

LOGGEROL_SEP = '/';

if ~exist(fname,'file')
  error('File not found: %s', fname)
end

fileID = fopen(fname);

line1 = fgetl(fileID);
if line1(1) == LOGGEROL_SEP
  % New style log
  newStyleLog = true;
  line1split = strsplit(line1(2:end), LOGGEROL_SEP);
  fprintf(1, 'New log style version %s\n', line1split{1});
  keys = strsplit(line1split{3},',');
  fmt = struct;%line1split{2};
  data = struct;
  for i=1:length(keys)
    [startIndex,endIndex] = regexp(keys{i}, '\d+');
    aa = keys{i};
    ndataincol = str2num(aa(startIndex:endIndex));
    data.(aa(1:startIndex-1)) = single(zeros(0, ndataincol));
    fmt.(aa(1:startIndex-1)) = aa(endIndex+1:end);
  end

  fprintf('Opened %s\nKeys: %s\n', fname, strjoin(fieldnames(data),','))
else
  % keep old
  newStyleLog = false;
  colnames = fgetl(fileID);
  fmt = fgetl(fileID);
  fprintf('Opened %s\nKeys: %s\nFormat: %s\n', fname, colnames, fmt)
  colnames = strsplit(colnames, ',');
  data = zeros([0,length(colnames)]);
end

i = 0;

try
  while ~feof(fileID)
    alignment1 = fread(fileID,1,'uint8');
    if alignment1 ~= 170
      continue;
    end
    alignment2 = fread(fileID,1,'uint8');
    if alignment2 ~= 187
      continue;
    end

    i = i+1;

    if newStyleLog
      fields = fieldnames(data);
      for j=1:numel(fields)
        colfmt = fmt.(fields{j});

        % fields{j}
        ndataincol = size(data.(fields{j}),2);
        if (colfmt=='f')
          type = 'float=>float';
        elseif (colfmt=='I')
          type = 'uint32=>float';
        elseif (colfmt =='B')
          type = 'uint8=>float';
        elseif (colfmt=='h')
          type = 'int16=>float';
        else
          error('unspecfied datatype in fmt');
        end
        newrow = fread(fileID,ndataincol,type)';
        if strcmp(fields{j},'eul') || strcmp(fields{j},'angVel') || strcmp(fields{j},'pos') || strcmp(fields{j},'torq') || strcmp(fields{j},'Vbat') || strcmp(fields{j},'xdot')
          newrow = 0.001 * newrow;
        elseif strcmp(fields{j},'acc') || strcmp(fields{j},'Ibat') 
          newrow = 0.01 * newrow;
        end
        data.(fields{j}) = [data.(fields{j}); newrow];
        % data
      end


    else
      % old same
      for j=1:length(fmt)
        if (fmt(j)=='f')
          type = 'float=>float';
        elseif (fmt(j)=='I')
          type = 'uint32=>float';
        elseif (fmt(j) =='B')
          type = 'uint8=>float';
        else
          error('unspecfied datatype in fmt');
        end

        data(i,j)=fread(fileID,1,type);
      end
      
    end
     
  end
  fclose(fileID);
catch
  disp('Finished reading');
end

if newStyleLog
  fprintf('Avg data rate = %.2f Hz\n',1000/mean(diff(data.t)))
  s = data;
else
  % Return data in a struct
  s = struct();
  for cn=1:length(colnames)
    % convert time to seconds
    if strcmp(colnames{cn},'t')
      data(:,cn) = 0.001 * data(:,cn);
    end
    s = setfield(s,colnames{cn},data(:,cn));
  end
end

end

File: ghost-sdk/tools/openlog/open_log.py
'''
 * Copyright (C) Ghost Robotics - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Avik De <avik@ghostrobotics.io>
'''
import time
from numpy import *
from pylab import *
import struct
# import sys
# import os, fnmatch
import argparse
import re

LOGGEROL_SEP = '/'

def readLogFile(filename, hasAlignmentWord = True, quiet = False):

	f = open(filename, 'rb')
	# if not quiet:
	print('Opened'),
	print(filename)

	# First two lines are special
	line1 = f.readline()
	newStyleLog = False
	if line1[0] == LOGGEROL_SEP:
		newStyleLog = True

	if newStyleLog:
		line1split = line1.lstrip(LOGGEROL_SEP).rstrip('\n').split(LOGGEROL_SEP)
		logVer = int(line1split[0])
		if not quiet:
			print 'New style log version', logVer

		keys = line1split[2].split(',')
		fmt = line1split[1]
		data = dict()
		ncols = len(keys)
		for i in range(ncols):
			# print keys[i]
			# figure out the keys and cols
			m = re.search(r'\d+', keys[i])
			ndataincol = int(m.group())
			# only keep the part before the int for key name
			keys[i] = keys[i][:m.start()]
			data[keys[i]] = zeros((0,ndataincol), dtype=np.float32)
	else:
		keys = line1.rstrip('\n').split(',')
		fmt = f.readline().rstrip('\n')
		data = dict.fromkeys(keys,zeros((0)))
		ncols = len(keys)

	sz = struct.calcsize(fmt)

	if not quiet:
		print('Keys:'),
		print(keys)
		print('Format:'),
		print(fmt)
		print('Size:'),
		print(sz)

	# print data
	# abort

	# Now just data
	if not hasAlignmentWord:
		line = f.read(sz)
		while len(line) == sz:
			tup = struct.unpack(fmt, line)

			for i in range(ncols):
				data[keys[i]] = r_[data[keys[i]],tup[i]]#vstack((data,array(tup)))
			line = f.read(sz)

	else:
		wholeFile = f.read()
		# split by alignment word
		chunks = wholeFile.split('\xaa\xbb')
		for chunk in chunks:
			# print len(chunk)
			# abort
			if len(chunk) == sz:
				# good chunk
				tup = struct.unpack(fmt, chunk)
				curpos = 0
				for i in range(ncols):
					ndataincol = data[keys[i]].shape[1]
					# print ndataincol
					if newStyleLog:
						newrow = array(tup[curpos:curpos+ndataincol])
						curpos = curpos + ndataincol
						# convert from compressed to float (opposite of LoggerOL)
						if keys[i] in ['eul','angVel','pos','torq','Vbat','xdot']:
							newrow = 0.001*newrow
						elif keys[i] in ['acc','Ibat']:
							newrow = 0.01*newrow
						data[keys[i]] = vstack((data[keys[i]], newrow))
						# print data[keys[i]]
						# abort
					else:
						data[keys[i]] = r_[data[keys[i]],tup[i]]

	if not quiet:
		print('Read data of length'),
		print(data[keys[0]].shape[0])
		if 't' in keys:
			if len(data['t'].shape) > 1:
				delt = diff(data['t'],axis=0)
			else:
				delt = diff(data['t'])
			print 'Avg data rate = {:2.2f} Hz'.format(1000/float(mean(delt)))

	return data

#
def trimData(data,inds):
	for key in data.keys():
		data[key] = data[key][inds]
	return data

def getRegions(tCol, stateCol = None):
	if stateCol is None:
		return None

	boundaries = r_[0, where(diff(stateCol) != 0)[0]]
	regions = zeros((0,3), dtype=int)

	if size(boundaries) > 1:
		for i in range(1,size(boundaries)):
			aa = (boundaries[i-1], boundaries[i])
			regions = vstack((regions, array([stateCol[boundaries[i]], aa[0], aa[1]], dtype=int)))
		# Add last region
		regions = vstack((regions, array([stateCol[boundaries[i]+1], aa[1], tCol.shape[0]-1], dtype=int)))
	return regions

def drawRegions(tCol, stateCol = None):
	if stateCol is None:
		return

	regions = getRegions(tCol, stateCol)
	cm = get_cmap('brg')

	for r in regions:
		bb = interp(r[0], r_[min(stateCol), max(stateCol)], r_[0,1])
		axvspan(tCol[r[1]], tCol[r[2]], facecolor=cm(bb), alpha=0.2, ec='none')

	xlim(min(tCol),max(tCol))

def drawVLines(tCol, every=None):  
	if every is None:
		return
	for x in arange(0,max(tCol),every):
		if x > min(tCol):
			axvline(x=x,ymin=-1000,ymax=1000,color='black',alpha=0.2)

def plotTrial(filename, trialType = None, trimRange = None, vlineEvery = None):
	beQuiet = (trialType == 'noplot')
	data = readLogFile(filename, hasAlignmentWord=True, quiet=beQuiet)
	if trimRange is not None:
		data = trimData(data, where((data['t'] >= trimRange[0]) & (data['t'] <= trimRange[1]))[0])

	if trialType == 'noplot':
		return data

	try:
		figure(int(filename[-9:-4]))
	except:
		figure()

	close('all')
	# Just plot all the data in rows
	# nrows = size(data.keys()) - 1
	nrows = 4
	currow = 1

	for key in data.keys():
		# For legibility open a new figure after 4
		if currow > 4:
			figure()
			currow = 1
		if key != 't' and key != 'st' and key != 'mo':
			subplot(nrows,1,currow)
			plot(0.001*data['t'], data[key])
			if 'mo' in data.keys():
				drawRegions(0.001*data['t'], data['mo'])
			if 'st' in data.keys():
				drawRegions(0.001*data['t'], data['st'])
			ylabel(key)
			currow = currow+1
			tight_layout()

	return data


#####################################################################################
''' SCRIPT '''

if __name__ == "__main__":

	parser = argparse.ArgumentParser(description='Plot some logs.')
	parser.add_argument('filename', help='filename to open (LOGxxxxx.TXT)')
	# parser.add_argument('-p', dest='pathRoot', help='Path under which to search for LOG*****.TXT files')
	parser.add_argument('-e', dest='trialType', help='experiment type {hop,freefall} (default: plot all data)')
	parser.add_argument('-t', nargs=2, dest='trimt', type=int, help='trim by time (default: show all)', metavar=('tstart', 'tend'))
	parser.add_argument('-v', dest='vlineEvery', type=int, help='vertical line every')
	args = parser.parse_args()

	data = plotTrial(args.filename, args.trialType, args.trimt, args.vlineEvery)
	show()

	# A = aEstimator()

